File: lib/api/github.ts
import { apiFetch } from "@/lib/api";

export interface GitHubRepository {
  id: string;
  fullName: string;
  private: boolean;
  htmlUrl: string;
  defaultBranch: string;
}

export interface MarkdownPublishResult {
  status: "published" | "up_to_date";
  path: string;
  commitSha: string | null;
  htmlUrl: string | null;
  publishedAt: string | null;
}

export interface AvailableGitHubRepository {
  githubRepositoryId: number;
  installationId: number;
  fullName: string;
  private: boolean;
}

export function getGitHubRepositories(): Promise<GitHubRepository[]> {
  return apiFetch<GitHubRepository[]>("/api/github/repositories");
}

export function getAvailableGitHubRepositories(): Promise<AvailableGitHubRepository[]> {
  return apiFetch<AvailableGitHubRepository[]>("/api/github/available-repositories");
}

export function connectGitHubRepository(repositoryId: number): Promise<GitHubRepository> {
  return apiFetch<GitHubRepository>("/api/github/repositories/connect", {
    method: "POST",
    body: JSON.stringify({ githubRepositoryId: repositoryId }),
  });
}

export function publishWorklog(
  worklogId: string,
  repositoryId: string,
  path: string,
  commitMessage: string,
): Promise<MarkdownPublishResult> {
  return apiFetch<MarkdownPublishResult>(`/api/worklogs/${encodeURIComponent(worklogId)}/publish`, {
    method: "POST",
    body: JSON.stringify({ repositoryId, path, commitMessage }),
  });
}

File: backend/app/schemas/github.py
from datetime import datetime
from uuid import UUID

from pydantic import Field

from app.schemas.worklog import APIModel


class GitHubRepositoryResponse(APIModel):
    id: UUID
    full_name: str
    private: bool
    html_url: str
    default_branch: str


class GitHubAvailableRepository(APIModel):
    github_repository_id: int
    installation_id: int
    full_name: str
    private: bool


class GitHubRepositoryConnect(APIModel):
    github_repository_id: int


class MarkdownPublishRequest(APIModel):
    repository_id: UUID
    path: str = Field(min_length=1, max_length=500)
    commit_message: str = Field(min_length=1, max_length=250)


class MarkdownPublishResponse(APIModel):
    status: str
    path: str
    commit_sha: str | None = None
    html_url: str | None = None
    published_at: datetime | None = None
